src/shmem_team.c

```diff
@@ -0,0 +1,455 @@
+
+    if (myteam->my_pe != -1) {
+
+        size_t psync = shmem_internal_team_choose_psync(parent_team, REDUCE);
+
+        shmem_internal_op_to_all(psync_pool_avail_reduced,
+                                 psync_pool_avail, N_PSYNC_BYTES, 1,
+                                 myteam->start, PE_stride, PE_size, NULL,
+                                 &shmem_internal_psync_pool[psync],
+                                 SHM_INTERNAL_BAND, SHM_INTERNAL_UCHAR);
+
+        shmem_internal_team_release_psyncs(parent_team, REDUCE);
+
+        /* Select the least signficant nonzero bit, which corresponds to an available pSync. */
+        myteam->psync_idx = shmem_internal_bit_1st_nonzero(psync_pool_avail_reduced, N_PSYNC_BYTES);
+        if (myteam->psync_idx == -1 || myteam->psync_idx >= shmem_internal_params.TEAMS_MAX) {
+            RAISE_WARN_MSG("No more teams available (max = %ld), try increasing SHMEM_TEAMS_MAX\n",
+                            shmem_internal_params.TEAMS_MAX);
+            myteam->psync_idx = -1;
+        } else {
+            /* Set the selected psync bit to 0, reserving that slot */
+            shmem_internal_bit_clear(psync_pool_avail, N_PSYNC_BYTES, myteam->psync_idx);
+
+            for (size_t i = 0; i < N_PSYNCS_PER_TEAM; i++)
+                myteam->psync_avail[i] = 1;
+
+            *new_team = myteam;
+
+            shmem_internal_team_pool[myteam->psync_idx] = *new_team;
+        }
+    }
+
+    size_t psync = shmem_internal_team_choose_psync(parent_team, SYNC);
+
+    shmem_internal_barrier(parent_team->start, parent_team->stride, parent_team->size,
+                           &shmem_internal_psync_barrier_pool[psync]);
+
+    shmem_internal_team_release_psyncs(parent_team, SYNC);
+
+    if (myteam->psync_idx == -1)
+        return -1;
+    else
+        return 0;
+}
+
+int shmem_internal_team_split_2d(shmem_internal_team_t *parent_team, int xrange,
+                                 const shmemx_team_config_t *xaxis_config, long xaxis_mask,
+                                 shmem_internal_team_t **xaxis_team, const shmemx_team_config_t *yaxis_config,
+                                 long yaxis_mask, shmem_internal_team_t **yaxis_team)
+{
+    const int parent_start = parent_team->start;
+    const int parent_stride = parent_team->stride;
+    const int parent_size = parent_team->size;
+    const int num_xteams = ceil( parent_size / (float)xrange );
+    const int num_yteams = xrange;
+
+    int start = parent_start;
+    int ret = 0;
+
+    for (int i = 0; i < num_xteams; i++) {
+	int xsize = (i == num_xteams - 1 && parent_size % xrange) ? parent_size % xrange : xrange;
+
+        if (shmem_internal_pe_in_active_set(shmem_internal_my_pe, start, parent_stride, xsize) != -1) {
+            ret = shmem_internal_team_split_strided(parent_team, start, parent_stride,
+                                            xsize, xaxis_config, xaxis_mask, xaxis_team);
```

Comment:
IIUC, this can also deadlock.  All of the PEs in the parent team need to call split_strided.